### PR TITLE
Fix parallel build issues

### PR DESCRIPTION
**Pull request by vehre**
[links]:#
[contributing guidelines]: https://github.com/sourceryinstitute/OpenCoarrays/blob/master/CONTRIBUTING.md
[issue]: https://github.com/sourceryinstitute/OpenCoarrays/issues
[coverage]: https://img.shields.io/codecov/c/github/sourceryinstitute/OpenCoarrays/master.svg?style=flat-square

|  coverage on master         |
|:---------------------------:|
| ![Codecov branch][coverage] |

## Summary of changes ##

Fix build dependencies to enable parallel builds.

## Rationale for changes ##

While figuring issues in OpenCoarrays fast builds are reQuired to reduce turn-around time when switching mpi-library. The build system was only allowing seQuential builds using make. The patches in this pr fix this for `make -j N` and `ninja`. Other build systems might benefit, too. This may allow for easier implementing #759, but will not fix it completely.

## Additional info and certifications ##

This pull request (PR) is a:

- [x] Bug fix
- [ ] Feature addition
- [ ] Other, Please describe:

### I certify that ###

- [x] I certify that:
  - I have reviewed and followed the [contributing guidelines]
  - I will wait at least 24 hours before self-approving the PR to give another
    OpenCoarrays developer a chance to review my proposed code
  - I have not introduced errant white space (no trailing white space or white space errors may
    be introduced)
  - I have added an explanation of what these changes do and why they should be included
  - I have checked to ensure there aren't other open [Pull Requests] for the same change
  - I have you written new tests for these changes
  - I have successfully tested these changes locally
  - I have commented any non-trivial, non-obvious code changes
  - The commits are logically atomic, self consistent and coherent
  - The [commit messages] follow [best practices]
  - Test coverage is maintained or increased after this is merged

## Code coverage data

coverage on master

[links]: #
[best practices]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[commit messages]: https://thoughtbot.com/blog/5-useful-tips-for-a-better-commit-message
[Pull Requests]: https://github.com/sourceryinstitue/OpenCoarrays/pulls

### CODE REVIEW

File: src/script-templates/caf.in [hunk 1]
```diff
@@ -161,14 +161,13 @@ __caf_err_report() {
...
-    for arg in "${@}"; do
-      if [[ "${arg}" == "-c" ]]; then
-        return 0
-      fi
-    done
-    return 1
-}
+with_link="true"
+for arg in "${@}"; do
+  if [[ "${arg}" == "-c" || "${arg}" == "-cpp" ]]; then
```
Comment (zbeekman) [inline src/script-templates/caf.in:166]: Good catch with `-cpp`.

File: src/script-templates/caf.in [hunk 2]
```diff
@@ -161,14 +161,13 @@ __caf_err_report() {
 trap '__caf_err_report "${FUNCNAME:-.}" ${LINENO}' ERR
 
 # See if we are compiling or compiling and/or linking
-__only_compiling () {
```
Comment (zbeekman) [inline src/script-templates/caf.in:164]: Just curious @vehre: what was your motivation for changing this from a function to a variable?

**Review by zbeekman — APPROVED** [review]
Looks good to me. I agree that these changes should prevent (at least some) parallel compilation failures. CMake is strange in how it names and generates `.mod` files and this should help to ensure that parallel builds aren't clobbering `.mod` files needed & in use by other targets. I want to double check what tests CI is running to ensure it includes the changed test cases.

**Comment by zbeekman** [comment]
> Looks good to me. I agree that these changes should prevent (at least some) parallel compilation failures. CMake is strange in how it names and generates `.mod` files and this should help to ensure that parallel builds aren't clobbering `.mod` files needed & in use by other targets. I want to double check what tests CI is running to ensure it includes the changed test cases.

I confirmed that this is running the tests that were changed in CI. Thanks for tackling this Andre! This is ready to merge AFAICT. Let me know if you want me to do the merge and whether or not to wait for @rouson to take a look.

**Comment by vehre** [comment]
@zbeekmann I switched from function to variable to evaluate it once only.
Else the function could possibly be evaluated multiple times with the same
result everytime. So just a small optimisation.

Am 8. Oktober 2024 20:00:09 schrieb zbeekman ***@***.***>:

**Review by rouson — APPROVED** [review]
LGTM

**Pull request merged**